plugins/modules/source_control/gitlab/gitlab_integration.py

```diff
@@ -0,0 +1,544 @@
+
+    def update(self, remote, active, params, events):
+        diff = not self.__equals(remote.attributes, active, params, events)
+        if diff:
+            # remote.active = active
+            for k, v in self.__expand_events(events).items():
+                setattr(remote, k, v)
+            # when saving, params does directly in the request
+            # (remember they get retrieved under the "property" key of the remote dict())
+            for k, v in params.items():
+                setattr(remote, k, v)
+            if not self._module.check_mode:
+                remote.save()
+            return remote
+
+        return False
+
+    def __as_api_object(self, active, params, events):
+        obj = {'active': active}
+        obj.update(self.__expand_events(events))
+        obj.update(params)
+        return obj
+
+    def __expand_events(self, events):
+        try:
+            # See https://gitlab.com/gitlab-org/gitlab/-/issues/27050 for why it's useful to
+            # discard unsupported events before comparing
+            supported_events = SRV_DEF[self.name]['_events']
+        except KeyError:
+            supported_events = None
+        ret = dict((value + '_events', bool(value in events)) for value in self.HOOK_EVENTS if supported_events is None or value in supported_events)
+        return ret
+
+    @staticmethod
+    def __events_equal(obj, events):
+        return all((k in obj and obj[k] == v) for k, v in events.items())
+
+    @staticmethod
+    def __has_credentials(attr):
+        return ('password' in attr and attr['password']) or ('token' in attr and attr['token']) or ('api_key' in attr and attr['api_key'])
+
+    def __equals(self, prev_attr, active, params, events):
+        filtered_params = dict((k, v) for k, v in params.items() if k not in self.CREDENTIAL_PARAMS)
+        if prev_attr['active'] != active:
+            self._module.debug("active status differs")
+            return False
+        if self.__has_credentials(params):
+            self._module.debug("has credentials: assuming difference")
+            return False
+        if not self.__events_equal(prev_attr, self.__expand_events(events)):
+            self._module.debug("events differs: %s != %s" % (prev_attr, events))
+            return False
+        if prev_attr['properties'] != filtered_params:
+            self._module.debug("integrations attributes differs: %s != %s" % (prev_attr['properties'], filtered_params))
+            return False
+        return True
+
+    def setattrs(self, obj, new_attr):
+        for k, v in new_attr.items():
+            setattr(obj, k, v)
+
+
+# Equivalent of doing `AnsibleModule(argument_spec=base_spec, **constraints)`
+def mimic_ansible_module_validation(specs, constraints):
+    from ansible.module_utils.basic import ModuleArgumentSpecValidator, _load_params
+    from ansible.module_utils.errors import UnsupportedError
+    import os
```

Comment:
These imports must be done on the global level, not locally in this function.

(This will also likely show that this does not work with Ansible 2.9 and ansible-base 2.10.)

Also you should not use `_load_params`. That's an internal API.

Finally, you should not re-validate the whole argument spec, but only the relevant option.

Reply:
- Ok about `global level function import`

> ` re-validate the whole argument spec, but only the relevant option`

But ... this implies, again, more code, more complexity than revalidating it (which does not hurt).


- About `you should not use  _load_params`, please please, have a look at the its inline documentation:
```
We cannot guarantee API stability for it (it may change between versions) however we will try not to break it gratuitously.  It is certainly more future-proof to call this function and consume its outputs than to implement the logic inside it as a copy in your own code.
```

=> The programmer who actually wrote this function tells us exactly the contrary : `if you need me, then use me`.

It feels bad how far fetched this debate about `AnsibleModule` is leading us.
My intent to do proper validation of arguments, of being strict and DRY ends up in more code, more workarounds, more complication, more maintenance hurdle, more duplication of the private code base because... well _the private method is private so lets copy the LoC instead of calling it_.

The documentation of `_load_params` is exactly telling us that : _Yes, I'm private but you know, it's better to reuse me than to rewrite me_ (and it's basically the same about the whole `AnsibleModule` which wasn't misused at all but **just instantiating twice** what **was never forbidden in the first place** ).

I feel we are loosing the big picture and the goal by blindly overfocusing on strict unquestioned "norms" (interpreted for the worst instead of the better) and that I'm pushed into a path that will lead to complex, bloated and inelegant code (that won't be merged either).

Reply:
> But ... this implies, again, more code, more complexity than revalidating it (which does not hurt).

It does not.

> * About `you should not use  _load_params`, please please, have a look at the its inline documentation:

It's still something that should really not be used in module code. It's ok to use in tests though.